### ISSUE TITLE
shufflenet different outputs than pytorch

### ISSUE THREAD

**Issue by makaveli10** [Tencent/ncnn#4816]
I have been benchmarking ncnn classification models
This is the script i used to validate if the outputs from ncnn match to the pytorch outputs with imagenet validation set:
```
import sys
import cv2
from ncnn.model_zoo import get_model
from ncnn.utils import print_topk
import numpy as np
import torch

model = torch.hub.load('pytorch/vision:v0.10.0', 'shufflenet_v2_x1_0', pretrained=True)

model.eval()

net = get_model("shufflenetv2", num_threads=4, use_gpu=False)


from PIL import Image
from torchvision import transforms


def torch_infer(img_path):
    input_image = Image.open(img_path)
    preprocess = transforms.Compose([
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        # transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
    ])
    input_tensor = preprocess(input_image)
    print(input_tensor.shape)
    input_batch = input_tensor.unsqueeze(0) # create a mini-batch as expected by the model

    # move the input and model to GPU for speed if available
    if torch.cuda.is_available():
        input_batch = input_batch.to('cuda')
        model.to('cuda')

    with torch.no_grad():
        output = model(input_batch)
    # Tensor of shape 1000, with confidence scores over Imagenet's 1000 classes
    # The output has unnormalized scores. To get probabilities, you can run a softmax on it.
    probabilities = torch.nn.functional.softmax(output[0], dim=0)
    return torch.argmax(probabilities)


def ncnn_infer(img_path):
    im = cv2.imread(img_path)
    print(im.shape)
    cls_score = net(im)

    return np.argmax(cls_score)



if __name__=="__main__":
    import os
    root_dir = '/workspace/ILSVRC/Data/CLS-LOC/val/'
    imgs = os.listdir(root_dir)
    count = 0
    for img in imgs[:1000]:
        
        img_path = os.path.join('/workspace/ILSVRC/Data/CLS-LOC/val', img)
        if Image.open(img_path).mode == "L": continue
        ncnn_out = ncnn_infer(img_path)
        torch_out = torch_infer(img_path).cpu().numpy()
        if ncnn_out != torch_out:
            print(torch_out, ncnn_out)
            print(img_path)
            count += 1
    print(count)

```
and out of 1000, 384 outputs were different.

**Comment by nihui** [comment]
The shufflenetv2 in ncnn model zoo is not converted from torchvision, but from the caffe model. Because the models are different, the results are different
You can convert torchvision shufflenet with the method here

https://github.com/Tencent/ncnn/issues/4681#issuecomment-1556868797

**Comment by makaveli10** [comment]
@nihui Thanks for the swift response, I got the model converted after building pnnx, but the problem is the normalization values, now I am doing inference on imagenet images like:
```
import ncnn
net = ncnn.Net()
net.opt.use_fp16_packed = False
net.opt.use_fp16_storage = False
net.opt.use_fp16_arithmetic = False
net.load_param("shufflenet_v2_x0_5.ncnn.param")
net.load_model("shufflenet_v2_x0_5.ncnn.bin")
ex = net.create_extractor()

input_image = cv2.imread(img_path)
mean_vals = []
norm_vals = [1 / 255.0, 1 / 255.0, 1 / 255.0]
mat_in = ncnn.Mat.from_pixels_resize(
    input_image,
    ncnn.Mat.PixelType.PIXEL_BGR,
    input_image.shape[1],
    input_image.shape[0],
    224,
    224,
    )
mat_in.substract_mean_normalize(mean_vals, norm_vals)

with net.create_extractor() as ex:
    ex.input("in0", mat_in)
    _, out0 = ex.extract("out0")
softmax = ncnn.create_layer("Softmax")
pd = ncnn.ParamDict()
softmax.load_param(pd)
softmax.forward_inplace(out0, net.opt)
mat_out = out0.reshape(out0.w * out0.h * out0.c)
cls_scores = np.array(mat_out)
np.argmax(cls_scores)
```
and I get 50 outputs out of 500 that match torch inference output with:
```
import torch
import torchvision.models as models
from PIL import Image
from torchvision import transforms

model = models.shufflenet_v2_x0_5(pretrained=True)
model.eval()

input_image = Image.open(img_path)
preprocess = transforms.Compose([
    transforms.Resize(256),
    transforms.CenterCrop(224),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
])
input_tensor = preprocess(input_image)
print(input_tensor)
input_batch = input_tensor.unsqueeze(0) # create a mini-batch as expected by the model

# move the input and model to GPU for speed if available
if torch.cuda.is_available():
    input_batch = input_batch.to('cuda')
    model.to('cuda')

with torch.no_grad():
    output = model(input_batch)
# Tensor of shape 1000, with confidence scores over Imagenet's 1000 classes
# The output has unnormalized scores. To get probabilities, you can run a softmax on it.
probabilities = torch.nn.functional.softmax(output[0], dim=0)
print(torch.argmax(probabilities))
```


the issue you mentioned for torch->ncnn conversion doesn't have the normalization params.

Any pointers on how to get the correct mean and norm vals?

**Comment by nihui** [comment]
https://github.com/Tencent/ncnn/wiki/FAQ-ncnn-produce-wrong-result#pre-process

**Comment by makaveli10** [comment]
@nihui thanks!
this is how my ncnn and pytorch validation-inference looks like now:
```
import sys
import cv2
from ncnn.model_zoo import get_model
from ncnn.utils import print_topk
import numpy as np
import torch
import torchvision.models as models

model = models.shufflenet_v2_x0_5(pretrained=True)
model.eval()

import ncnn
net = ncnn.Net()
net.opt.use_fp16_packed = False
net.opt.use_fp16_storage = False
net.opt.use_fp16_arithmetic = False
net.load_param("shufflenet_v2_x0_5.ncnn.param")
net.load_model("shufflenet_v2_x0_5.ncnn.bin")

def ncnn_infer(img_path):
    input_image = cv2.imread(img_path)
    mean_vals = [0.485*255.0 , 0.456*255.0, 0.406*255.0]
    norm_vals = [1/0.229/255.0, 1/0.224/255.0, 1/0.225/255.0]
    mat_in = ncnn.Mat.from_pixels_resize(
        input_image,
        ncnn.Mat.PixelType.PIXEL_BGR,
        input_image.shape[1],
        input_image.shape[0],
        224,
        224,
    )
    mat_in.substract_mean_normalize(mean_vals, norm_vals)
    # print(mat_in)
    with net.create_extractor() as ex:
        ex.input("in0", mat_in)
        _, out0 = ex.extract("out0")
        softmax = ncnn.create_layer("Softmax")
        pd = ncnn.ParamDict()
        softmax.load_param(pd)
        softmax.forward_inplace(out0, net.opt)
        mat_out = out0.reshape(out0.w * out0.h * out0.c)
        cls_scores = np.array(mat_out)
    return np.argmax(cls_scores)

def torch_infer(img_path):
    input_image = Image.open(img_path)
    preprocess = transforms.Compose([
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
    ])
    input_tensor = preprocess(input_image)
    input_batch = input_tensor.unsqueeze(0) # create a mini-batch as expected by the model

    # move the input and model to GPU for speed if available
    if torch.cuda.is_available():
        input_batch = input_batch.to('cuda')
        model.to('cuda')
    # print(input_batch)
    with torch.no_grad():
        output = model(input_batch)
        # print(output)

    # Tensor of shape 1000, with confidence scores over Imagenet's 1000 classes
    # The output has unnormalized scores. To get probabilities, you can run a softmax on it.
    probabilities = torch.nn.functional.softmax(output[0], dim=0)
    print(torch.argmax(probabilities) )
    return torch.argmax(probabilities) 
```

I think I am missing something, because out 497 images I get the class that matches torch shufflenet for 181 images.
